### DISCUSSION TITLE
SNMP v3 - adding to LibreNMS

### DISCUSSION THREAD

**Discussion by monteroman** [bonzo81/netbox-librenms-plugin#27] [Q&A] [answered]
Hey there.

I am trying to get a switch that's in LibreNMS as a SNMP v3 switch to show up in NetBox's LibreNMS tab, but it doesn't seem to see it. How does one go about adding the sync for the switch ports if it is set up already in LibreNMS?  I tried just putting in a bogus v2 SNMP string, but it just says that the switch was already added.

Thanks much!

**Comment by bonzo81** [comment]
Hi @monteroman ,

Check that the hostname (IP) in LibreNMS matches the primary IP for the device in Netbox. That's the most likely reason the plugin can't find your device in LibreNMS.

The SNMP option only shows if the Netbox device has an primary IP but it can't find a match in LibreNMS. If you know the device is in LibreNMS then you disregard that form.

Hope that helps.

**Reply by monteroman** [reply]
So is it looking for a host name or an IP address in LibreNMS? 
And is it case sensitive?

**Reply by monteroman** [reply]
How would I then go about syncing the interfaces to Netbox if it's not available in that tab/form?

**Reply by bonzo81** [reply]
Hi,

At the moment the plugin uses IP to lookup LibreNMS. So if the device is added to LibreNMS using a FQDN, it won't find it.

It is on the roadmap to use hostname for lookup too, but it's not added yet I'm afraid.

The environment I use to build the plugin doesn't use FQDN for LibreNMS devices, but I think I can find a way to test this and build it into the next release.

**Reply by monteroman** [reply]
I am looking forward to the FQDN matching. We have several hundred entries in LibreNMS and it's all on FQDN. 

I temporarily renamed the hostname to IP in LNMS and it did find it immediately in Netbox. 
Thank you for your quick response!

**Reply by bonzo81** [reply]
Hi,

Out of interest, do you put the FQDN as the device name in Netbox, or just the hostname?

I'm now looking into auto discovering the LibreNMS device_id, and using that for API calls to LMNS. 

Thanks for the feedback

**Comment by monteroman** [comment]
We use FQDN in Netbox.On Nov 16, 2024, at 10:15 AM, Andy Norwood ***@***.***> wrote:﻿
Hi,
Out of interest, do you put the FQDN as the device name in Netbox, or just the hostname?
I'm now looking into auto discovering the LibreNMS device_id, and using that for API calls to LMNS.
Thanks for the feedback

**Reply by monteroman** [reply]
So I was giving some thought to this. 
In Netbox, some of our engineers are using the hostname only and not FQDN, but we are doing FQDN in LibreNMS.
Would there be a way to regex the hostname when checking if the host is in LNMS?

**Reply by bonzo81** [reply]
I have been working on this too.

The limiting factor, as I understand it, would be the LibreNMS API endpoint. Any device related query accepts either hostname (FQDN or IP depending on how the device was added to LIbreNMS) or ID as the identifier. Anything else would result in an error. The API doc is here https://docs.librenms.org/API/Devices/#get_device

Currently, I’m testing an approach that involves storing the LibreNMS device ID in NetBox as a custom field. Then using it as the primary method to query the LibreNMS API for data. This setup requires users to create a custom field called `librenms_id` for devices in NetBox. This my preferred approach as it makes the calls to the LibreNMS API predictable and consistent.

The plugin will try to populate the `librenms_id` field automatically by attempting to use the  primary IP or NetBox device name (whether FQDN or not) in the API call. The success of the API call is determined by the available data in NetBox.

Of course, the `librenms_id` field can also be populated manually. Which would be a possible solution for your devices that don't have the FQDN in NetBox. But it might be preferable to update the hostname to FQDN rather then add value to the librenms_id custom field! If the `librenms_id` custom field does not exist, the plugin can still function, but again relying on the same lookup logic as described above.

That's where I'm at with it at the moment. Still testing.

**Reply by monteroman** [reply]
The 0.2.9 version is working well, but I have a request. Can you make the hostname case insensitive when it refers to LNMS for a lookup? I've found at times that some of hour hostnames are upper or mixed case in Netbox and LNMS can be either upper or lower so it throws a mismatch and won't sync until one is renamed to match the case. Even if you're using the LNMS device#.

Thanks!

**Reply by bonzo81** [reply, marked as answer]
Thanks for the feedback. I've not been able to replicate this in my environment but I can see where this could happen in the code. I'll apply a fix, and it should be in the next 0.3.0 release in a couple of days.

**Reply by monteroman** [reply]
Bonzo, it looks like 0.3.1 fixed the issue with the case-sensitive host names.  Thank you!

**Reply by bonzo81** [reply]
Glad to hear that!
